src/uct/ib/dc/dc_mlx5.c

```diff
@@ -224,15 +224,25 @@ uct_dc_mlx5_poll_tx(uct_dc_mlx5_iface_t *iface)
     ucs_trace_poll("dc iface %p tx_cqe: dci[%d] qpn 0x%x txqp %p hw_ci %d",
                    iface, dci, qp_num, txqp, hw_ci);
 
+    /* Process completions before updating resources to avoid OOO.
+     * Some resources (RDMA_READ credits) are updated in operation completion
+     * callbacks, so need to call all relevant callbacks prior dispatching
+     * arbiter. In this case ordering is preserved, because:
+     * - If nothing is sent from the completion callback, we just update
+     *   resources and dispatch arbiter.
+     * - If something needs to be sent from completion callback, but not all
+     *   needed resources available, it will be added to the pending queue
+     *   (and handled in dispatch below in the proper order).
```

Comment:
what if we have some pending ops because no read resources, then 1st send completion op releases read resources, and 2nd completion sends an active message because resources were just released?
it will reorder w.r.t whatever is on the pending queue already.